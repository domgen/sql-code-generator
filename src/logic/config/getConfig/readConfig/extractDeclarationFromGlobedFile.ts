import { QueryDeclaration, ResourceDeclaration } from '../../../../domain';
import { extractSqlFromFile } from '../../../common/extractSqlFromFile';

export enum DeclarationType {
  QUERY = 'QUERY',
  RESOURCE = 'RESOURCE",',
}

export const extractDeclarationFromGlobedFile = async ({
  rootDir,
  relativePath,
  type,
}: {
  rootDir: string;
  relativePath: string;
  type: DeclarationType;
}) => {
  // 0. define the file path
  const filePath = `${rootDir}/${relativePath}`;

  // 1. validate the file type
  const sql = await extractSqlFromFile({ filePath });

  // 3. define the declaration
  if (type === DeclarationType.QUERY) {
    return new QueryDeclaration({
      path: relativePath,
      sql,
    });
  }
  if (type === DeclarationType.RESOURCE) {
    return new ResourceDeclaration({
      path: relativePath,
      sql,
    });
  }
  throw new Error('unexpected'); // fail fast
};
