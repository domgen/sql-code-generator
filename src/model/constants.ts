export enum DatabaseLanguage {
  MYSQL = 'mysql',
}
export enum DefinitionType {
  RESOURCE = 'resource',
  QUERY = 'query',
}
