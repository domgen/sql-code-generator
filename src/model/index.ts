export * from './constants';
export * from './valueObjects';
