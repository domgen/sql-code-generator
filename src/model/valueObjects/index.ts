export * from './GeneratorConfig';
export * from './QueryDefinition';
export * from './ResourceDefinition';
