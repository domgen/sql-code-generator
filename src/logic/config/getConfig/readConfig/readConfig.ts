import { readYmlFile } from './utils/readYmlFile';
import { GeneratorConfig } from '../../../../model';
import { getAllPathsMatchingGlobs } from '../getAllPathsMatchingGlobs/getAllPathsMatchingGlobs';
import { extractResourceDeclarationFromGlobedFile } from '../extractDeclarationFromGlobbedFile/extractResourceDeclarationFromGlobedFile';
import { extractQueryDeclarationFromGlobedFile } from '../extractDeclarationFromGlobbedFile/extractQueryDeclarationFromGlobedFile';

/*
  1. read the yml file
  2. get the connection (i.e., import from the specified connection file)
  3. get the definitions and flatten them
*/
export const readConfig = async ({ filePath }: { filePath: string }) => {
  const configDir = filePath
    .split('/')
    .slice(0, -1)
    .join('/'); // drops the file name

  // get the yml
  const contents = await readYmlFile({ filePath });

  // get the output dir
  if (!contents.generates) throw new Error('generates key must be defined');
  if (!contents.generates.types) throw new Error('generates.types must specify where to output the generated types');
  if (!contents.generates.queryFunctions) {
    throw new Error('generates.queryFunctions must specify where to output the query functions');
  }

  // get the language and dialect
  if (!contents.language) throw new Error('language must be defined');
  const language = contents.language;
  if (!contents.dialect) throw new Error('dialect must be defined');
  const dialect = `${contents.dialect}`; // ensure that we read it as a string, as it could be a number

  // get the resource declarations
  const resourceGlobs = contents.resources;
  const resourcePaths = await getAllPathsMatchingGlobs({
    globs: resourceGlobs,
    root: configDir,
  });
  const resourceDeclarations = await Promise.all(
    resourcePaths
      .sort() // for determinism in order
      .map((relativePath) => extractResourceDeclarationFromGlobedFile({ rootDir: configDir, relativePath })),
  );

  // get the query declarations
  const queryGlobs = contents.queries;
  const queryPaths = await getAllPathsMatchingGlobs({
    globs: queryGlobs,
    root: configDir,
  });
  const queryDeclarations = await Promise.all(
    queryPaths
      .sort() // for determinism in order
      .map((relativePath) => extractQueryDeclarationFromGlobedFile({ rootDir: configDir, relativePath })),
  );

  // return the results
  return new GeneratorConfig({
    rootDir: configDir,
    generates: {
      types: contents.generates.types,
      queryFunctions: contents.generates.queryFunctions,
    },
    language,
    dialect,
    declarations: [...resourceDeclarations, ...queryDeclarations],
  });
};
