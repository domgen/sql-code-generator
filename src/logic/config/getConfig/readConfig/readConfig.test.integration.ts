import { GeneratorConfig } from '../../../../model';
import { readConfig } from './readConfig';

describe('readConfig', () => {
  it('should be able to read the example config provisioned in __test_assets__', async () => {
    const config = await readConfig({ filePath: `${__dirname}/__test_assets__/sql.yml` });
    expect(config).toBeInstanceOf(GeneratorConfig);
    expect(config.language).toEqual('mysql');
    expect(config.dialect).toEqual('5.7');
    expect(config.definitions.length).toEqual(4); // 4 total: 2 queries, 2 resources
    expect(config).toMatchSnapshot(); // to log an example of the output
  });
});
