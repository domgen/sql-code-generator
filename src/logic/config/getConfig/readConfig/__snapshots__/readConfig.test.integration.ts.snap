// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`readConfig should be able to read the example config provisioned in __test_assets__ 1`] = `
GeneratorConfig {
  "definitions": Array [
    QueryDefinition {
      "name": "selectOne",
      "sql": "select 1",
    },
    ResourceDefinition {
      "name": "communication_channel",
      "sql": "-- ----------------------------------------
-- create communication channel table
-- ----------------------------------------
CREATE TABLE \`communication_channel\` (
  -- meta
  \`id\` BIGINT NOT NULL AUTO_INCREMENT PRIMARY KEY,
  \`uuid\` VARCHAR(36) NOT NULL,
  \`created_at\` DATETIME DEFAULT CURRENT_TIMESTAMP,

  -- static data
  \`type\` ENUM('APP', 'SMS', 'EMAIL') NOT NULL, -- e.g., \\"how do we send this\\"
  \`address\` VARCHAR(36) NOT NULL, -- e.g., \\"where do we send this\\"

  -- meta meta
  UNIQUE (\`type\`, \`address\`) -- type + address uniquely identify a communication channel
) ENGINE = InnoDB;
",
      "type": "TABLE",
    },
    ResourceDefinition {
      "name": "upsert_user_description",
      "sql": "/*
  this is an example procedure with an example comment at the top, just to prove that we can extrac thte name and type from it
*/
CREATE PROCEDURE upsert_user_description(
  IN in_from_user_id BIGINT,
  IN in_first_name VARCHAR(255),
  IN in_last_name VARCHAR(255),
  IN in_phone_channel_id BIGINT
)
BEGIN
  -- ----------------------------------------
  -- DECLARATIONS
  -- ----------------------------------------
  DECLARE v_data_hash BINARY(20);
  DECLARE v_entity_exists BOOLEAN DEFAULT false;

  -- -----------------------------------------
  -- UPSERT RECORD
  -- -----------------------------------------
  -- calculate the hash
  SET v_data_hash = UNHEX( --  get the binary
    SHA( -- of the hex hash
      CONCAT_WS( -- of a string combining all non-null values
        ':',
        in_from_user_id,
        in_first_name,
        in_last_name,
        in_phone_channel_id
      )
    )
  );

  -- ... do more stuff
END;
",
      "type": "PROCEDURE",
    },
    QueryDefinition {
      "name": "findAllUsersByName",
      "sql": "SELECT * FROM users WHERE name = :name",
    },
  ],
  "dialect": "5.7",
  "language": "mysql",
}
`;
